internal/tracinginterceptor/interceptor.go

```diff
@@ -39,32 +44,76 @@ var (
 
 // Params defines the parameters for creating the Interceptor
 type Params struct {
-	// Tracer is used to propagate context and to generate spans
-	Tracer opentracing.Tracer
-	// Transport is the name of the transport, it decides the propagation format and propagation carrier
+	Tracer    opentracing.Tracer
 	Transport string
 }
 
 // Interceptor is the tracing interceptor for all RPC types.
 // It handles both tracing observability and context propagation using OpenTracing APIs.
 type Interceptor struct {
+	tracer            opentracing.Tracer
+	transport         string
+	propagationFormat opentracing.BuiltinFormat
 }
 
 // New constructs a tracing interceptor with the provided parameter.
 func New(p Params) *Interceptor {
-	return &Interceptor{}
+	m := &Interceptor{
+		tracer:            p.Tracer,
+		transport:         p.Transport,
+		propagationFormat: tchannel.GetPropagationFormat(p.Transport),
+	}
+	if m.tracer == nil {
+		m.tracer = opentracing.GlobalTracer()
+	}
+	return m
 }
 
 // Handle implements interceptor.UnaryInbound
 func (m *Interceptor) Handle(ctx context.Context, req *transport.Request, resw transport.ResponseWriter, h transport.UnaryHandler) error {
-	// TODO: implement
-	panic("implement me")
+	parentSpanCtx, _ := m.tracer.Extract(m.propagationFormat, tchannel.GetPropagationCarrier(req.Headers.Items(), req.Transport))
+	tags := ExtractTracingTags()
+
+	extractOpenTracingSpan := &transport.ExtractOpenTracingSpan{
+		ParentSpanContext: parentSpanCtx,
+		Tracer:            m.tracer,
+		TransportName:     req.Transport,
+		StartTime:         time.Now(),
+		ExtraTags:         tags,
+	}
+	ctx, span := extractOpenTracingSpan.Do(ctx, req)
+	defer span.Finish()
+
+	err := h.Handle(ctx, req, resw)
+	return updateSpanWithError(span, err)
```

Comment:
For inbound/ handler, checking the error is probably not enough - for TChannel, there are failure cases where error is nil. And we probably need similar [writer wrapper implementation](https://github.com/yarpc/yarpc-go/blob/dev/internal/observability/middleware.go#L40) in observability middleware.

Reply:
It doesn't seem to be resolved.

Reply:
So, I see writer wrapper was copied here.

1. Does it mean, we'll have two wrappers now for the same request? So, one wrapper calls another wrapper that calls real handler?
2. Is there a plan to remove existing one? I guess, there are no?
3. As a side note: response size, for example, is not even in use, why was it copied here?

So, I guess we need to have one writer wrapper, it should be one layer of this header (which should be reused by obs and tracing).

Reply:
For now, we plan to implement two wrappers to minimize the workload. In the future, we aim to consolidate them into one.